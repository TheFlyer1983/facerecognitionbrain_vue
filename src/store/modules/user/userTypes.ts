import { ActionContext } from 'vuex';

export type UserActionContext = ActionContext<UserState, {}>

export interface UserState {
  isSignedIn: boolean;
  isProfileOpen: false;
  user: User | null;
  token: string;
  rank: string;
}

export interface User {
  id: string;
  name: string;
  email: string;
  entries: number;
  joined: string;
  pet: string;
  age: string;
}
