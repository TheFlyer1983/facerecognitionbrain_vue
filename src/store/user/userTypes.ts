export interface UserState {
  isSignedIn: boolean;
  isProfileOpen: boolean;
  id: string;
  user: User | null;
  token: string;
  rank: string;
}

export interface User {
  name: string;
  email: string;
  entries: number;
  joined: string | number | Date;
  pet?: string;
  age?: number;
  role?: string;
}

export interface UpdateInfo {
  name?: string;
  pet?: string;
  age?: number;
}
