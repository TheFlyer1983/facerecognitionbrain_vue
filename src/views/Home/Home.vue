<script setup lang="ts">
import { computed, onBeforeMount } from 'vue';
import { useStore } from 'vuex';
import Logo from '@/components/Logo/Logo.vue'

const store = useStore();

onBeforeMount(() => {
  if (!store.getters['user/getUser']) {
    store.dispatch('user/getToken');
  }
});

const user = computed(() => store.getters['user/getUser']);
</script>

<template>
  <Logo />
  <h1>{{ user }}</h1>
</template>

<style scoped>
a {
  color: #42b983;
}

label {
  margin: 0 0.5em;
  font-weight: bold;
}

code {
  background-color: #eee;
  padding: 2px 4px;
  border-radius: 4px;
  color: #304455;
}
</style>
