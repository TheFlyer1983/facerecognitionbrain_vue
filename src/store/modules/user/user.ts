import { UserState, UserActionContext } from './userTypes';

import { endpoints } from '@/constants';
import { LoginInfo, RegisterInfo } from '@/types';
import {
  getAuthTokenInSession,
  saveAuthTokenInSession
} from '@/functions/storageFunctions';

export const state: UserState = {
  isSignedIn: false,
  isProfileOpen: false,
  user: null,
  token: ''
};

export const getters = {
  getUser(state: UserState): UserState['user'] {
    return state.user;
  },
  getToken(state: UserState): UserState['token'] {
    return state.token;
  },
  getIsSignedIn(state: UserState): UserState['isSignedIn'] {
    return state.isSignedIn;
  }
};

export const mutations = {
  setUserId(state: UserState, payload: string): void {
    state.user = { ...state.user, id: payload } as UserState['user'];
  },

  setUser(state: UserState, payload: UserState['user']): void {
    state.user = payload;
  },

  toggleSignIn(state: UserState, payload: UserState['isSignedIn']): void {
    state.isSignedIn = payload;
  },

  setToken(state: UserState, payload: UserState['token']): void {
    state.token = payload;
  }
};

export const actions = {
  async login(
    { commit, dispatch }: UserActionContext,
    payload: LoginInfo
  ): Promise<boolean> {
    try {
      const response = await fetch(endpoints.signin, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload)
      });

      const result = await response.json();

      if (!response.ok) throw new Error(result as string);

      saveAuthTokenInSession(result.token as string);
      commit('setToken', result.token);
      commit('setUserId', result.userId);

      const success = await dispatch('getUser', result.userId as string);
      if (!success) throw new Error('Error');

      return true;
    } catch (error) {
      console.log(error, 'In the error block');
      return false;
    }
  },
  async getUser({ commit }: UserActionContext, payload: string) {
    const requestURL = endpoints.getProfile.replace(':id', payload);
    const token = getAuthTokenInSession();
    try {
      if (!token) throw new Error('No valid token');
      const response = await fetch(requestURL, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          Authorization: token
        }
      });
      const result = await response.json();
      if (!response.ok) throw new Error(result);

      commit('setUser', result);
      commit('toggleSignIn', true);
      return true;
    } catch (error) {
      console.log(error);
      return false;
    }
  },
  async registerUser(
    { commit }: UserActionContext,
    payload: RegisterInfo
  ): Promise<boolean> {
    try {
      const response = await fetch(endpoints.register, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload)
      });

      const result = await response.json();

      if (!response.ok) throw new Error(result as string);

      saveAuthTokenInSession(result.session.token as string);
      commit('setUser', result.register.user);

      return true;
    } catch (error) {
      console.log(error);
      return false;
    }
  },
  getToken({ commit, dispatch }: UserActionContext) {
    const token = getAuthTokenInSession();

    if (token) {
      commit('setToken', token);
      dispatch('authenticated');
    }
  },
  async authenticated({ state, dispatch }: UserActionContext) {
    const token = state.token;

    if (token) {
      try {
        const response = await fetch(endpoints.signin, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            Authorization: token
          }
        });

        const result = await response.json();

        if (!response.ok) throw new Error(result as string);

        dispatch('getUser', result.id);
      } catch (error) {
        console.log(error);
      }
    }
  }
};

const userModule = {
  namespaced: true,
  state,
  getters,
  mutations,
  actions
};

export default userModule;
