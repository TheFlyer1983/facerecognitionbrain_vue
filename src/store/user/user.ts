import { defineStore } from 'pinia';
import { UpdateInfo, User, UserState } from './userTypes';
import {
  LoginInfo,
  LoginResponse,
  RegisterInfo,
  RegisterResponse
} from '@/types';
import request from '@/functions/request';
import { endpoints } from '@/constants';
import {
  getAuthTokenInSession,
  removeAuthTokenFromSession,
  saveAuthTokenInSession
} from '@/functions/storageFunctions';

export const useUserStore = defineStore('user', {
  state: (): UserState => ({
    isSignedIn: false,
    isProfileOpen: false,
    id: '',
    user: null,
    token: '',
    rank: ''
  }),

  actions: {
    async login(payload: LoginInfo) {
      payload = { ...payload, returnSecureToken: true };

      try {
        const response = await request.post<LoginResponse>(
          endpoints.signin,
          payload,
          {
            params: { key: import.meta.env.VITE_APP_FIREBASE_API_KEY }
          }
        );

        saveAuthTokenInSession(response.data.idToken);
        this.token = response.data.idToken;
        this.id = response.data.localId;

        const success = await this.getUser(response.data.localId);
        if (!success) throw new Error('Error');

        return true;
      } catch (error) {
        console.error(error);
        return false;
      }
    },

    async getUser(userId: string) {
      const requestURL = endpoints.profile.replace(':id', userId);

      try {
        const response = await request.get<User>(requestURL, {
          params: { auth: this.token }
        });

        this.user = response.data;
        this.isSignedIn = true;

        //this.getRank();

        return true;
      } catch (error) {
        console.error(error);
        return false;
      }
    },

    async registerUser(payload: RegisterInfo) {
      const registerPayload = {
        email: payload.email,
        password: payload.password,
        returnSecureToken: true
      };
      try {
        const response = await request.post<RegisterResponse>(
          endpoints.register,
          registerPayload,
          {
            params: { key: import.meta.env.VITE_APP_FIREBASE_API_KEY }
          }
        );

        saveAuthTokenInSession(response.data.idToken);
        this.token = response.data.idToken;
        this.id = response.data.localId;
        await this.createProfile(payload.name);

        return true;
      } catch (error) {
        console.error(error);
        return false;
      }
    },

    async createProfile(name: string) {
      const profilePayload = {
        name,
        entries: 0,
        joined: new Date()
      };

      const requestURL = endpoints.profile.replace(':id', this.id);
      try {
        const response = await request.put(requestURL, profilePayload, {
          params: { auth: this.token }
        });

        this.user = response.data;
      } catch (error) {
        console.log(error);
      }
    },

    getToken() {
      const token = getAuthTokenInSession();

      if (token) {
        this.token = token;

        this.authenticated();
      }
    },

    async authenticated() {
      if (this.token) {
        try {
          const response = await request.post(endpoints.signin);

          this.getUser(response.data.id);
        } catch (error) {
          console.error(error);
        }
      }
    },

    async getRank() {
      const entries = this.user?.entries;

      try {
        const response = await request.post(endpoints.rank, { entries });
        this.rank = response.data;
      } catch (error) {
        console.error(error);
      }
    },

    signout() {
      removeAuthTokenFromSession();
      this.$reset();
    },

    async updateUser(payload: UpdateInfo) {
      const requestURL = endpoints.profile.replace(':id', this.id);
      try {
        await request.put(requestURL, payload);

        this.getUser(this.id as string);
      } catch (error) {
        console.error(error);
      }
    },

    async deleteUser() {
      const requestURL = endpoints.profile.replace(':id', this.id);
      try {
        await request.delete(requestURL, { params: { auth: this.token } });

        await request.post(
          endpoints.delete,
          { idToken: this.id },
          { params: { key: import.meta.env.VITE_APP_FIREBASE_API_KEY } }
        );

        this.signout();
      } catch (error) {
        console.error(error);
      }
    }
  }
});
