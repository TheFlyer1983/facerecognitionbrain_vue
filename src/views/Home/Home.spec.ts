import { describe, expect, it, beforeEach, vi, beforeAll } from 'vitest';
import { shallowMount, VueWrapper } from '@vue/test-utils';
import { reactive } from 'vue';
import { useStore } from 'vuex';

import Home from './Home.vue';
import { UserMock } from '@/fixtures/users';
import { User } from '@/store/modules/user/userTypes';


vi.mock('vuex');

const mockedUseStore = vi.mocked<() => Partial<typeof useStore>>(useStore);

describe('Given the `Home` component', () => {
  const render = () =>
    shallowMount(Home, {
      global: {
        stubs: {
          teleport: true
        }
      }
    });
  let wrapper: VueWrapper<InstanceType<typeof Home>>;

  const mockedGetters = {
    ['user/getUser']: UserMock as User | null,
    ['user/getIsProfileOpen']: false
  };

  const mockStore = reactive({
    getters: mockedGetters,
    dispatch: vi.fn()
  });

  beforeAll(() => {
    mockedUseStore.mockReturnValue(mockStore);
  });

  describe('when it is rendered', () => {
    beforeEach(() => {
      wrapper = render();
    });

    it('should match the snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    describe('when there is no current user loaded', () => {
      beforeEach(() => {
        mockStore.getters['user/getUser'] = null;

        wrapper = render();
      });

      it('should dispatch the correct action', () => {
        expect(mockStore.dispatch).toHaveBeenCalledWith('user/getToken');
      });
    });

    describe('when the profile modal is open', () => {
      beforeEach(() => {
        mockStore.getters['user/getIsProfileOpen'] = true;
      });

      it('should match the snapshot', () => {
        expect(wrapper.element).toMatchSnapshot();
      });
    });
  });
});
