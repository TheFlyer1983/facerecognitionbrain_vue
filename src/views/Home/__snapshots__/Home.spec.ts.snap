// Vitest Snapshot v1

exports[`Given the \`Home\` component > when it is rendered > should match the snapshot 1`] = `
<div
  data-v-app=""
>
  
  <logo-stub />
  <rank-stub
    user="[object Object]"
  />
  <image-link-form-stub />
  <face-recognition-stub />
  <teleport-stub>
    <!--v-if-->
  </teleport-stub>
  
</div>
`;

exports[`Given the \`Home\` component > when it is rendered > when the profile modal is open > should match the snapshot 1`] = `
<div
  data-v-app=""
>
  
  <logo-stub />
  <!--v-if-->
  <image-link-form-stub />
  <face-recognition-stub />
  <teleport-stub>
    <profile-stub />
  </teleport-stub>
  
</div>
`;
